From 7c17adce87436606c98715204363f139ea7c88dd Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?C=C3=A9dric=20Bosdonnat?= <cedricbosdo@openoffice.org>
Date: Thu, 27 Jan 2011 17:01:02 +0100
Subject: [PATCH] Allow to add funky email addresses as the first field of an alias

---
 ConfigFile.py |    8 ++++----
 1 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/ConfigFile.py b/ConfigFile.py
index 4b7a9ce..041289e 100644
--- a/ConfigFile.py
+++ b/ConfigFile.py
@@ -43,12 +43,12 @@ def ReadEmailAliases (name):
         croak ('Unable to open email alias file %s' % (name))
     line = ReadConfigLine (file)
     while line:
-        sline = line.split ()
-        if len (sline) != 2:
+        m = re.match ('^("[^"]+"|\S+)\s+(.+)$', line)
+        if not m or len (m.groups ()) != 2:
             croak ('Funky email alias line "%s"' % (line))
-        if sline[0].find ('@') <= 0 or sline[1].find ('@') <= 0:
+        if m and m.group (2).find ('@') <= 0:
             croak ('Non-addresses in email alias "%s"' % (line))
-        database.AddEmailAlias (sline[0], sline[1])
+        database.AddEmailAlias (m.group (1).replace ('"', ''), m.group (2))
         line = ReadConfigLine (file)
     file.close ()
 
-- 
1.7.2.2

