From 9c6170d8d62112dd600025b98d9cd87faedf9f86 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?C=C3=A9dric=20Bosdonnat?= <cedricbosdo@openoffice.org>
Date: Mon, 14 Feb 2011 11:28:24 +0100
Subject: [PATCH 2/3] Made the CSV file aggregating data by weeks or months

---
 csv.py |    4 +++-
 gitdm  |   11 ++++++++---
 2 files changed, 11 insertions(+), 4 deletions(-)

diff --git a/csv.py b/csv.py
index 34ea10a..e97e620 100644
--- a/csv.py
+++ b/csv.py
@@ -15,8 +15,10 @@ class CSVStat:
 
 PeriodCommitHash = { }
 
-def AccumulatePatch (p):
+def AccumulatePatch (p, Aggregate):
     date = "%.2d-%.2d-01"%(p.date.year, p.date.month)
+    if (Aggregate == 'week'):
+        date = "%.2d-%.2d"%(p.date.year, p.date.isocalendar()[1])
     authdatekey = "%s-%s"%(p.author.name, date)
     if authdatekey not in PeriodCommitHash:
         empl = p.author.emailemployer (p.email, p.date)
diff --git a/gitdm b/gitdm
index 47d1966..6dd6e7e 100755
--- a/gitdm
+++ b/gitdm
@@ -36,6 +36,7 @@ AkpmOverLt = 0
 DumpDB = 0
 CFName = 'gitdm.config'
 DirName = ''
+Aggregate = 'month'
 
 #
 # Options:
@@ -52,14 +53,16 @@ DirName = ''
 # -s		Ignore author SOB lines
 # -u		Map unknown employers to '(Unknown)'
 # -x file.csv   Export raw statistics as CSV
+# -A level  Aggregrate the raw statistics according to level (week or month).
+#           The default value is "month"
 # -z		Dump out the hacker database at completion
 
 def ParseOpts ():
     global MapUnknown, DevReports
     global DateStats, AuthorSOBs, FileFilter, AkpmOverLt, DumpDB
-    global CFName, CSVFile, DirName
+    global CFName, CSVFile, DirName, Aggregate
 
-    opts, rest = getopt.getopt (sys.argv[1:], 'ab:dc:Dh:l:o:r:sux:z')
+    opts, rest = getopt.getopt (sys.argv[1:], 'aA:b:dc:Dh:l:o:r:sux:z')
     for opt in opts:
         if opt[0] == '-a':
             AkpmOverLt = 1
@@ -87,6 +90,8 @@ def ParseOpts ():
         elif opt[0] == '-x':
             CSVFile = open (opt[1], 'w')
             print "open output file " + opt[1] + "\n"
+        elif opt [0] == '-A':
+            Aggregate = opt[1]
         elif opt[0] == '-z':
             DumpDB = 1
         
@@ -370,7 +375,7 @@ while (1):
         for hacker in p.reports:
             hacker.addreport (p)
     CSCount += 1
-    csv.AccumulatePatch (p)
+    csv.AccumulatePatch (p, Aggregate)
 print >> sys.stderr, 'Grabbing changesets...done       '
 
 if DumpDB:
-- 
1.7.2.2

