From f37d3417041f6c58f8183f3f1ccb7ddfd12624f8 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?C=C3=A9dric=20Bosdonnat?= <cedricbosdo@openoffice.org>
Date: Thu, 21 Oct 2010 00:46:58 +0200
Subject: [PATCH] Only add Linus and Andrew with -a

---
 gitdm |    8 +++++---
 1 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/gitdm b/gitdm
index dce5d3c..022a73a 100755
--- a/gitdm
+++ b/gitdm
@@ -295,7 +295,8 @@ def ApplyFileFilter (line, ignore):
 # remove the (redundant) Linus signoff.
 #
 def TrimLTSOBs (p):
-    if Linus in p.sobs and Akpm in p.sobs:
+    global AkpmOverLt
+    if AkpmOverLt == 1 and Linus in p.sobs and Akpm in p.sobs:
         p.sobs.remove (Linus)
 
 
@@ -313,9 +314,10 @@ ConfigFile.ConfigFile (CFName, DirName)
 # Let's pre-seed the database with a couple of hackers
 # we want to remember.
 #
-Linus = ('torvalds@linux-foundation.org',
+if AkpmOverLt == 1:
+    Linus = ('torvalds@linux-foundation.org',
          LookupStoreHacker ('Linus Torvalds', 'torvalds@linux-foundation.org'))
-Akpm = ('akpm@linux-foundation.org',
+    Akpm = ('akpm@linux-foundation.org',
         LookupStoreHacker ('Andrew Morton', 'akpm@linux-foundation.org'))
 
 NextLine = sys.stdin.readline ()
-- 
1.6.0.2

