From c148a3ad6103cbb5ddcfb5eaa6858421ad85b9ee Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?C=C3=A9dric=20Bosdonnat?= <cedricbosdo@openoffice.org>
Date: Mon, 25 Oct 2010 16:40:38 +0200
Subject: [PATCH 5/5] Made the CSV file aggregating data by weeks

---
 csv.py |   10 +++++-----
 1 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/csv.py b/csv.py
index 34ea10a..94db89c 100644
--- a/csv.py
+++ b/csv.py
@@ -16,11 +16,11 @@ class CSVStat:
 PeriodCommitHash = { }
 
 def AccumulatePatch (p):
-    date = "%.2d-%.2d-01"%(p.date.year, p.date.month)
+    date = "%.2d-%.2d"%(p.date.year, p.date.isocalendar()[1])
     authdatekey = "%s-%s"%(p.author.name, date)
     if authdatekey not in PeriodCommitHash:
         empl = p.author.emailemployer (p.email, p.date)
-        stat = CSVStat (p.author.name, empl, date)
+        stat = CSVStat (p.author.name, empl, p.date)
         PeriodCommitHash[authdatekey] = stat
     else:
         stat = PeriodCommitHash[authdatekey]
@@ -29,10 +29,10 @@ def AccumulatePatch (p):
 def OutputCSV (file):
     if file is None:
         return
-    file.write ("Name\tAffliation\tDate\tAdded\tRemoved\n")
+    file.write ("Name\tAffliation\tYear\tWeek\tAdded\tRemoved\n")
     for date, stat in PeriodCommitHash.items():
         # sanitise names " is common and \" sometimes too
         empl_name = stat.employer.name.replace ("\"", ".").replace ("\\", ".")
         author_name = stat.name.replace ("\"", ".").replace ("\\", ".")
-        file.write ("\"%s\"\t\"%s\"\t%s\t%d\t%d\n"%(author_name, empl_name, stat.date, \
-                                                    stat.added, stat.removed))
+        file.write ("\"%s\"\t\"%s\"\t%d\t%d\t%d\t%d\n"%(author_name, empl_name, stat.date.year, \
+                                                    stat.date.isocalendar()[1], stat.added, stat.removed))
-- 
1.7.2.2

